src/compat/db_json.cpp

```diff
@@ -0,0 +1,933 @@
+{
+  assert (doc->IsObject());
+
+  doc->AddMember (rapidjson::StringRef (name), JSON_VALUE ().SetInt (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_member_to_object (JSON_DOC *doc, char *name, JSON_DOC *value)
+{
+  assert (doc->IsObject());
+
+  if (value->IsArray())
+    {
+      doc->AddMember (rapidjson::StringRef (name), value->GetArray(), doc->GetAllocator ());
+    }
+  else if (value->IsObject())
+    {
+      doc->AddMember (rapidjson::StringRef (name), value->GetObject(), doc->GetAllocator ());
+    }
+}
+
+void
+db_json_add_member_to_object (JSON_DOC *doc, char *name, float value)
+{
+  assert (doc->IsObject());
+
+  doc->AddMember (rapidjson::StringRef (name), JSON_VALUE ().SetFloat (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_member_to_object (JSON_DOC *doc, char *name, double value)
+{
+  assert (doc->IsObject());
+
+  doc->AddMember (rapidjson::StringRef (name), JSON_VALUE ().SetDouble (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_element_to_array (JSON_DOC *doc, char *value)
+{
+  if (!doc->IsArray())
+    {
+      doc->SetArray();
+    }
+
+  doc->PushBack (rapidjson::StringRef (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_element_to_array (JSON_DOC *doc, int value)
+{
+  if (!doc->IsArray())
+    {
+      doc->SetArray();
+    }
+
+  doc->PushBack (JSON_VALUE ().SetInt (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_element_to_array (JSON_DOC *doc, float value)
+{
+  if (!doc->IsArray())
+    {
+      doc->SetArray();
+    }
+
+  doc->PushBack (JSON_VALUE ().SetFloat (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_element_to_array (JSON_DOC *doc, double value)
+{
+  if (!doc->IsArray())
+    {
+      doc->SetArray();
+    }
+
+  doc->PushBack (JSON_VALUE ().SetDouble (value), doc->GetAllocator ());
+}
+
+void
+db_json_add_element_to_array (JSON_DOC *doc, JSON_DOC *value)
+{
+  if (!doc->IsArray())
+    {
+      doc->SetArray();
+    }
+
+  doc->PushBack (*value, doc->GetAllocator ());
+}
+
+JSON_DOC *
+db_json_get_json_from_str (const char *json_raw, int &error_code)
+{
+  JSON_DOC *new_doc = new JSON_DOC;
+
+  if (new_doc->Parse (json_raw).HasParseError())
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_INVALID_JSON, 2,
+              rapidjson::GetParseError_En (new_doc->GetParseError ()), new_doc->GetErrorOffset ());
+      delete new_doc;
+      error_code = ER_INVALID_JSON;
+      return NULL;
+    }
+
+  error_code = NO_ERROR;
+  return new_doc;
+}
+
+JSON_DOC *
+db_json_get_copy_of_doc (const JSON_DOC *doc)
+{
+  JSON_DOC *new_doc = new JSON_DOC;
+  new_doc->CopyFrom (*doc, new_doc->GetAllocator ());
+
+  return new_doc;
+}
+
+void
+db_json_copy_doc (JSON_DOC *dest, JSON_DOC *src)
+{
+  dest->CopyFrom (*src, dest->GetAllocator());
+}
+
+void
+db_json_insert_func (JSON_DOC *doc, char *raw_path, char *str_value, int &error_code)
+{
+  JSON_POINTER p (raw_path);
+  JSON_DOC inserting_doc;
+  JSON_DOC aux;
+
+  aux.CopyFrom (*doc, aux.GetAllocator ());
```

Comment:
Avoid CopyFrom and memory duplication as much as possible.

Reply:
this can't be avoided because JSON_POINTER::Set does a shallow copy and inserting_doc will be freed, this leading to memory corruption